src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -461,14 +465,57 @@ fn all_constructors<'a, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
                 .map(|v| Variant(v.did))
                 .collect()
         }
+        ty::TyChar if exhaustive_integer_patterns => {
+            let (min, max) = (0u128, char::MAX as u128);
+            value_constructors = true;
+            vec![ConstantRange(ty::Const::from_bits(cx.tcx, min, cx.tcx.types.char),
+                               ty::Const::from_bits(cx.tcx, max, cx.tcx.types.char),
+                               RangeEnd::Included)]
+        }
+        ty::TyInt(int_ty) if exhaustive_integer_patterns => {
+            use syntax::ast::IntTy::*;
+            macro_rules! min_max_ty {
+                ($ity:ident, $uty:ty, $sty:expr) => {
+                    ($ity::MIN as $uty as u128, $ity::MAX as $uty as u128, $sty)
+                }
+            }
+            let (min, max, ty) = match int_ty {
+                Isize => min_max_ty!(isize, usize, cx.tcx.types.isize),
+                I8    => min_max_ty!(i8, u8, cx.tcx.types.i8),
+                I16   => min_max_ty!(i16, u16, cx.tcx.types.i16),
+                I32   => min_max_ty!(i32, u32, cx.tcx.types.i32),
+                I64   => min_max_ty!(i64, u64, cx.tcx.types.i64),
+                I128  => min_max_ty!(i128, u128, cx.tcx.types.i128),
```

Comment:
cc @oli-obk This sort of thing can be done from the bit width (which you can get from the layout of these types).